### PR TITLE
Collecting All Validation Errors on .validate() Instead of Raising Exception

### PR DESCRIPTION
**Pull request by cerivera**
Addresses: https://github.com/spotify/ramlfications/issues/21

New usage:

    errors = ramlfications.validate(RAML_FILE)
    if errors:
        for error in errors:
            print(error)
    else:
        print("Success!")

`ramlfications.parse` also collects all errors, but raises an exception for the first validation error in the `errors` list;  this was to keep the new functionality close to how it works now.  

Implementation Notes:
- Initialized an `errors` array in `parser.py` that gets passed through the node creation -- similar to `config`.
- Created `parse_raml_passive` that sits next to `parse_raml` so that `ramlfications.parse` and `ramlfications.validate` can have the different behaviors that are described above;  `parse_raml_passive` does not raise an exception after parsing, and `parse_raml` does.

### CODE REVIEW

**Comment by econchick** [comment]
Hey @cerivera - super sorry for the delay!  I was traveling when you submitted this, but I'm back now and will review/comment today/tomorrow.  Thanks for your patience!

File: .gitignore [hunk 1]
```diff
@@ -29,3 +29,4 @@ coverage.xml
 docs/_build/
 docs/_templates/
 cover
+.idea/
```
Comment (econchick) [inline .gitignore]: if you could actually just add this to your git exclude file, that'd be great (since it's specific to *your* set up).  e.g. within the main `ramlfications` directory, `$ echo ".idea" >> .git/exclude`

File: docs/usage.rst [hunk 2]
```diff
@@ -95,19 +95,23 @@ To validate a RAML file with Python:
...
+   >>> if not errors:
+   >>>     print("Success!")
 
 .. code-block:: python
 
    >>> from ramlfications import validate
    >>> RAML_FILE = "/path/to/invalid/no-title.raml"
-   >>> validate(RAML_FILE)
-   InvalidRootNodeError: RAML File does not define an API title.
+   >>> errors = validate(RAML_FILE)
```
Comment (econchick) [inline docs/usage.rst]: More commenting on the implemented behavior rather than the docs:

This breaks the behavior of the current API of the package.

Currently, the `validate` function returns nothing if the file is valid, and the (first) error if it's not:

```pyc
>>> RAML_FILE = "invalid.raml"
>>> validate(RAML_FILE)
InvalidParameterError: '422' not a valid HTTP response code.
>>> RAML_FILE = "valid.raml"
>>> validate(RAML_FILE)
>>>
```

However, with this, it's now:

```pyc
>>> RAML_FILE = "invalid.raml"
>>> validate(RAML_FILE)
[ramlfications.errors.InvalidParameterError("'422' not a valid HTTP response code.")]
>>> RAML_FILE = "valid.raml"
>>> validate(RAML_FILE)
[]
>>>
```

The proposed implementation returns a list in either case, the original returns nothing if successful, and an error message if invalid.   I'd like to at least keep the "return nothing if valid" behavior.  To be honest, not sure what I'd like to have for invalid. But when originally implemented, I wanted to be able to do something like this:

```python
try:
    validate(RAML_FILE)
except Exception:
    raise SomeError
```

So with the proposed implementation, the `try` will succeed and not hit the `except` clause.

**Comment by econchick** [comment]
Hey @cerivera!

So I think here may be a better way to do this.

First, there's no need to have an empty `return` statement within the `validators.py`.  Python will automatically do that for you.

I'm wondering if rather than rewrite all the `validators.py` code to "collect" the errors in a list, that there could be a decorator to collect everything.  Something like:

```
@collecterrors
def root_version(inst, attr, value):
    ...
```

What do you think?

File: .gitignore (hunk 1 above)
Comment (cerivera) [inline .gitignore]: Will do!

File: docs/usage.rst (hunk 2 above)
Comment (cerivera) [inline docs/usage.rst]: I agree that a successful validation should return nothing instead of an empty list.  I'll put that back in.  

Hrmm, if we revert to raising an exception on invalid, how do you propose we report ALL errors that were collected?  Would we create a new Exception class that knows how to work with multiple errors instead of just raising an individual error?

**Comment by cerivera** [comment]
I like that decorator idea a lot!  I'll work on it.

File: docs/usage.rst (hunk 2 above)
Comment (econchick) [inline docs/usage.rst]: Yeah I think that may be the approach here, have something like `MultipleValidationErrors(Exception)`.

**Comment by econchick** [comment]
:+1: thanks so much!

**Comment by econchick** [comment]
Hey @cerivera  - any update on this PR?  Please ping me here if you need a review.

**Comment by cerivera** [comment]
@econchick I plan on getting to it this weekend; i got pulled into some stuff right after our last conversation.

**Comment by econchick** [comment]
@Cerivera awesome! Not a problem and don't mean to rush you at all. It's just a really great feature that I'm excited to have a contributor for :) let me know if I can help in any way!

**Comment by cerivera** [comment]
@econchick Fixes are in!

Here's the new usage (copied from usage.rst):

    >>> from ramlfications import validate
    >>> RAML_FILE = "/path/to/my-api.raml"
    >>> validate(RAML_FILE)
    >>>


    >>> from ramlfications import validate
    >>> RAML_FILE = "/path/to/invalid/no-title.raml"
    >>> validate(RAML_FILE)
    InvalidRootNodeError: RAML File does not define an API title.


    >>> from ramlfications import validate
    >>> RAML_FILE = "/path/to/invalid/no-base-uri-no-title.raml"
    >>> validate(RAML_FILE)
    MultipleValidationErrors:
    InvalidRootNodeError: RAML File does not define the baseUri.
    InvalidRootNodeError: RAML File does not define an API title.

**Comment by econchick** [comment]
Hey @cerivera !  Thanks so much!!  I haven't looked at the code yet, but the usage you wrote looks like the exact behavior that I imagine. 

I'm needing to make a flight home to Seattle for the week (family emergency) but I should have this reviewed in the next day or two.

File: docs/usage.rst [hunk 3]
```diff
@@ -105,9 +106,18 @@ To validate a RAML file with Python:
...
+   >>> RAML_FILE = "/path/to/invalid/no-base-uri-no-title.raml"
+   >>> validate(RAML_FILE)
+   MultipleValidationErrors:
+   InvalidRootNodeError: RAML File does not define the baseUri.
+   InvalidRootNodeError: RAML File does not define an API title.
+
 .. note::
     When using ``validate`` within Python (versus the command line utility), if the RAML \
-    file is valid, then nothing is returned.  Only invalid files will return an exception.
+    file is valid, then nothing is returned; only invalid files will return an exception.  If multiple errors are found, a ``MultipleValidationErrors`` exception is raised, and you can access the invididual errors via the ``errors`` attribute.
```
Comment (econchick) [inline docs/usage.rst]: nitpick: typo in "individual"

File: ramlfications/decorators.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+from .errors import InvalidRAMLError
```
Comment (econchick) [inline ramlfications/decorators.py]: How would you feel making this module semiprivate? e.g. `ramlfications/_decorators.py` to signal it's not exactly a public API?

File: ramlfications/parser.py [hunk 5]
```diff
@@ -27,20 +28,35 @@
 def parse_raml(loaded_raml, config):
     """
     Parse loaded RAML file into RAML/Python objects.
+    Raise exception on validation errors.
```
Comment (econchick) [inline ramlfications/parser.py]: I think this should be moved down a couple of lines after "returns", like 

```
:raises: :py:class:`.errors.MultipleValidationErrors` or :py:class:`.errors.InvalidRAMLError` when RAML file is invalid
```

**Comment by econchick** [comment]
hey @cerivera ! nice job - the code looks good to me!  Just a couple of things that I already mentioned that I'd like your thoughts on.  Also if you could squash the 20 commits into one or two commits (I like a clean git history), it would be awesome.

Thanks for the great work!

**Comment by cerivera** [comment]
@econchick thanks for the feedback!  i agreed with all points and pushed a new commit that overwrote the rest.

File: tests/data/validate/optional-base-uri-params.raml [hunk 6]
```diff
@@ -24,11 +24,7 @@ uriParameters:
       example: "foo-api"
   get:
     description: "Get some foo"
-    type: "string"
```
Comment (econchick) [inline tests/data/validate/optional-base-uri-params.raml]: curious - why were these lines removed?

File: docs/usage.rst [hunk 7]
```diff
@@ -105,9 +106,21 @@ To validate a RAML file with Python:
    >>> validate(RAML_FILE)
    InvalidRootNodeError: RAML File does not define an API title.
 
+.. code-block:: python
+
+   >>> from ramlfications import validate
+   >>> RAML_FILE = "/path/to/invalid/no-base-uri-no-title.raml"
+   >>> validate(RAML_FILE)
+   MultipleValidationErrors:
```
Comment (econchick) [inline docs/usage.rst]: So I was thinking/re-reviewing this morning (the review a couple of days ago was while traveling and not fully/mentally "present" :smile:).

Correct me if I'm wrong - but the following would happen:

If there is only *one* error within the RAML file, this patch would still raise `MultipleValidationErrors`.

How would you feel renaming the `MultipleValidationErrors` to `InvalidRAML`?  To avoid the incorrect pluralization (e.g. `you have 1 errors`).

Comment (econchick) [inline docs/usage.rst]: (continuing thought) - you'd have to probably rename the current `InvalidRAMLError` to something like `BaseRAMLError`.

File: tests/data/validate/optional-base-uri-params.raml (hunk 6 above)
Comment (cerivera) [inline tests/data/validate/optional-base-uri-params.raml]: This was throwing a `MultipleValidationErrors` exception because you're defining a type on an HTTP method.  It was unrelated to the test, in my opinion.

File: ramlfications/errors.py [hunk 8]
```diff
@@ -31,6 +31,19 @@ class InvalidSecuritySchemeError(InvalidRAMLError):
...
+class MultipleValidationErrors(Exception):
+    def __init__(self, errors):
+        super(MultipleValidationErrors, self).__init__(
+            "Multiple validation errors were found.")
+        self.errors = errors
+
+    def __str__(self):
+        output = "\n"
+        for e in self.errors:
+            output += "{0}: {1}\n".format(e.__class__.__name__, e)
```
Comment (econchick) [inline ramlfications/errors.py]: What do you thinking of adding a `\t` (tab) when printing, at least to the console?

```python
MultipleValidationErrors:
    InvalidRootNodeError: RAML File does not define the baseUri.
    InvalidRootNodeError: RAML File does not define an API title.
```

File: tests/data/validate/optional-base-uri-params.raml (hunk 6 above)
Comment (econchick) [inline tests/data/validate/optional-base-uri-params.raml]: ahhhh gotcha! thanks!

File: docs/usage.rst (hunk 7 above)
Comment (cerivera) [inline docs/usage.rst]: At the end of `parse_raml` in `parser.py`, I'm only raising a `MultipleValidationErrors` exception if there are more than one:

    if root.errors:
        if len(root.errors) == 1:
            raise root.errors[0]
        else:
            raise MultipleValidationErrors(root.errors)

Comment (econchick) [inline docs/usage.rst]: AH right okay 

/me goes to get more coffee

**Comment by econchick** [comment]
Ok so pardon my stupid comments/questions this morning :)

Everything looks good!  Merging now - will include this in the 1.7 release which will go out in the next day or so.

File: docs/usage.rst (hunk 7 above)
Comment (cerivera) [inline docs/usage.rst]: :) hrmm now i'm wondering if it should have a consistent behavior regardless of number of errors...

Comment (econchick) [inline docs/usage.rst]: haven't merged yet - if you'd like!

(also just realized the `\t` comment below - lmk what you think, I'll hold off)

Comment (cerivera) [inline docs/usage.rst]: Hrmm, if we changed the behavior of the validation exceptions, we're going to have to update all of the tests in `test_validate.py` to look for the error in `errors`.

Comment (econchick) [inline docs/usage.rst]: You're right - it's up to you either way.

**Comment by cerivera** [comment]
@econchick I replaced `MultipleValidationErrors` with `InvalidRAMLError` and tabbed the errors in the `__str__` method.

**Comment by econchick** [comment]
@cerivera if you're ever in SF, let me buy you a beer :)

Thanks for everything - this looks really good to me. As I said before - will include this in the next release (1.7) in the next couple of days (just want to wait for #40 )

**Comment by cerivera** [comment]
@econchick woohoo!  I learned a ton working on your project, and I do like beer :)

ciao!
